### PR TITLE
CLDR-16763 Reduce deadlocks

### PR DESCRIPTION
**Pull request by srl295**
CLDR-16763

- [ ] This PR completes the ticket.

Initial work:

- use a real cache instead of homegrown
- initialize the non-resolved XMLSource / CLDRFile in the Per-Locale c'tor instead of deferring

I think initializing the resolved XMLSource might be more challenging because it's reentrant, needs to construct other PerLocaleDatas. 

But perhaps, when, say, French or French in Canada is hit, ST should grab a lock  and initialize ALL `fr*` sublocales right there. That way if  say `fr_FR` and `fr_CA` both got hit at the same time, they wouldn't deadlock on adding `fr` . 

ALLOW_MANY_COMMITS=true

### CODE REVIEW

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/STFactory.java [hunk 1]
```diff
@@ -1650,10 +1667,20 @@ public BallotBox<User> ballotBoxForLocale(CLDRLocale locale) {
     }
 
     /** Per locale map */
-    private final Map<CLDRLocale, Reference<PerLocaleData>> locales = new HashMap<>();
-
-    private final Cache<CLDRLocale, PerLocaleData> rLocales =
-            CacheBuilder.newBuilder().softValues().build();
+    private final LoadingCache<CLDRLocale, PerLocaleData> locales =
+            CacheBuilder.newBuilder()
+                    .concurrencyLevel(5)
```
Comment (srl295) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/STFactory.java:1672]: OK, i should set softValues  and perhaps some duration timeout.

**Comment by srl295** [comment]
this is looking promising. 
Needs a longer cache time. But the reentrancy seems to be fixed.

**Review by macchiati — DISMISSED** [review]
Looks good to me.

**Review by btangmu — DISMISSED** [review]
I looked through the changes quickly and no problems were obvious to me, however, this is a huge amount of change and I didn't understand a lot of this code in the first place, so basically a rubber-stamp LGTM

**Comment by srl295** [comment]
> I looked through the changes quickly and no problems were obvious to me, however, this is a huge amount of change and I didn't understand a lot of this code in the first place, so basically a rubber-stamp LGTM

It's hopefully more understandable with smaller units and more comments.

**Comment by srl295** [comment]
just one more commit since approval, https://github.com/unicode-org/cldr/pull/3047/commits/d5b9870d42c69c90fc03fd75134d48e798bb4b92 for productization

Pushing this to cldr-staging

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/STFactory.java [hunk 2]
```diff
@@ -908,6 +691,11 @@ public VoteResolver<String> getResolver(String path) {
             return getResolver(peekXpathData(path), path, null);
         }
 
+        @Override
+        public VoteResolver<String> getResolver(String path, VoteResolver<String> r) {
+            return getResolver(peekXpathData(path), path, r);
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/STFactory.java:696]: minor quibble: I prefer to put caller before callee -- easier to read with higher-level methods higher in the file

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyMain.java [hunk 3]
```diff
@@ -2516,36 +2511,19 @@ public void close() {
...
-                resolvedFile = cldrfile;
-            }
         }
 
-        public void clear() {
-            // TODO: try just kicking these instead of clearing?
-            cldrfile = null;
-            dbSource = null;
-        }
+        public void clear() {}
```
Comment (btangmu) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyMain.java:2526]: is `clear()` still called anywhere? could just remove it?

**Review by btangmu — APPROVED** [review]
lgtm though a lot of it is still mysterious to me

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyMain.java (hunk 3 above)
Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyMain.java:2526]: I'm not sure, I didn't add it in this PR

Comment (btangmu) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyMain.java:2526]: no callers, not even from .jsp

Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/SurveyMain.java:2526]: I think UserLocaleStuff was for convenience of JSPs and other very old flows… to make it easier to get the 'current' CLDRFile.  Probably most of the class is obsolete.

**Comment by srl295** [comment]
@btangmu PTAL. Let's discuss merging tomorrow if no issues show up

**Review by btangmu — APPROVED** [review]

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/LocaleMaxSizer.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.TreeMap;
+import java.util.TreeSet;
+import org.unicode.cldr.util.CLDRLocale;
+
+/**
+ * This class tracks the expected maximum size of strings in the locale.
+ *
+ * @author srl
+ */
+public class LocaleMaxSizer {
```
Comment (macchiati) [inline tools/cldr-apps/src/main/java/org/unicode/cldr/web/LocaleMaxSizer.java:14]: Question. By maximum size, do you mean the maximum **length** of any element value in the main locale file? It looks like the code assumes that, and also assumes that the longest one is the exemplar value in the CJK locales which is 8192, and that every other locale has a max of 4096. And we don't care that it is not exact numbers. If these are true, would be nice to add to the javadoc. (but not a blocker)

**Review by macchiati — APPROVED** [review]

File: tools/cldr-apps/src/main/java/org/unicode/cldr/web/LocaleMaxSizer.java (hunk 4 above)
Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/LocaleMaxSizer.java:14]: Yes length

Comment (srl295) [reply tools/cldr-apps/src/main/java/org/unicode/cldr/web/LocaleMaxSizer.java:14]: Created a ticket https://unicode-org.atlassian.net/browse/CLDR-16777 to document.

**Pull request merged**